src/Config/Factory/Cache.php

```diff
@@ -183,21 +211,82 @@ private function getSlaveConnection(): array
-            if ((isset($envCacheConfig['frontend'][$type]['backend_options']['server'])
-                    && $envCacheConfig['frontend'][$type]['backend_options']['server'] !== $redisConfig['host'])
-                || (isset($envCacheConfig['frontend'][$type]['backend_options']['port'])
-                    && $envCacheConfig['frontend'][$type]['backend_options']['port'] !== $redisConfig['port'])
-            ) {
+        if ($this->isNewConfigStructure()) {
+            $host = $envCacheConfig['frontend']['default']['backend_options']['remote_backend_options']['server']
+                ?? null;
+
+            $port = $envCacheConfig['frontend']['default']['backend_options']['remote_backend_options']['port']
+                ?? null;
+
+            if (($host !== null && $host !== $redisConfig['host'])
+                || ($port !== null && $port !== $redisConfig['port'])) {
                 return false;
             }
+        } else {
+            foreach (['default', 'page_cache'] as $type) {
+                $host = $envCacheConfig['frontend'][$type]['backend_options']['server'] ?? null;
+                $port = $envCacheConfig['frontend'][$type]['backend_options']['port'] ?? null;
+
+                if (($host !== null && $host !== $redisConfig['host'])
+                    || ($port !== null && $port !== $redisConfig['port'])) {
+                    return false;
+                }
+            }
         }
 
         return true;
     }
+
+    private function getOldConfigStructure(string $envCacheBackendModel, array $redisConfig): array
+    {
+        return [
+            'backend' => addslashes($envCacheBackendModel),
+            'backend_options' => [
+                'server' => $redisConfig['host'],
+                'port' => $redisConfig['port'],
+            ]
+        ];
+    }
+
+    private function getNewConfigStructure(string $envCacheBackendModel, array $redisConfig): array
+    {
+        return [
+            'backend' => addslashes($envCacheBackendModel),
+            'backend_options' => [
+                'remote_backend' => addslashes('\Magento\Framework\Cache\Backend\Redis'),
+                'remote_backend_options' => [
+                    'server' => $redisConfig['host'],
+                    'port' => $redisConfig['port'],
+                    'database' => self::REDIS_DATABASE_DEFAULT,
+                    'persistent' => 0,
+                    'password' => '',
+                    'compress_data' => '1',
+                ],
+                'local_backend' => 'Cm_Cache_Backend_File',
+                'local_backend_options' => [
+                    'cache_dir' => '/dev/shm/'
+                ]
+            ],
+            'frontend_options' => [
+                'write_control' => false,
+            ]
+        ];
+    }
+
+    /**
+     * @return bool
+     * @throws \Magento\MagentoCloud\Config\ConfigException
+     */
+    private function isNewConfigStructure(): bool
```

Comment:
"new" and "old" are sort of vague and don't really scale well. Perhaps something more descriptive like "synchronized cache" and...? "unsynced cache"?